source/lac/sparse_direct.cc

```diff
@@ -838,6 +838,200 @@ SparseDirectUMFPACK::n() const
 }
 
 
+
+#ifdef DEAL_II_WITH_MUMPS
+
+SparseDirectMUMPS::SparseDirectMUMPS()
+  : initialize_called(false)
+{}
+
+SparseDirectMUMPS::~SparseDirectMUMPS()
+{
+  id.job = -2;
+  dmumps_c(&id);
+
+  // Do some cleaning
+  if (Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0)
+    {
+      delete[] a;
+      delete[] irn;
+      delete[] jcn;
+    }
+}
+
+template <class Matrix>
+void
+SparseDirectMUMPS::initialize_matrix(const Matrix &matrix)
+{
+  Assert(matrix.n() == matrix.m(), ExcMessage("Matrix needs to be square."));
+
+
+  // Check we haven't been here before:
+  Assert(initialize_called == false, ExcInitializeAlreadyCalled());
+
+  // Initialize MUMPS instance:
+  id.job = -1;
+  id.par = 1;
+  id.sym = 0;
+
+  // Use MPI_COMM_WORLD as communicator
+  id.comm_fortran = -987654;
```

Comment:
What were we thinking here?

Reply:
This is actually correct (though someone who uses this interface should document this and get rid of the `new`s, etc): the MUMPS manual states

> or the C interface, MUMPS requires a Fortran communicator to be provided in id.comm fortran. If, however, this field is initialized to the special value -987654, the Fortran communicator `MPI_COMM_WORLD` is used by default.

See circa page 114 of https://sources.debian.org/data/main/m/mumps/5.7.3-3/doc/userguide_5.7.3.pdf.

Reply:
Well, that's both funny and pretty bizarre. Only Fortran and C programmers tend to come up with these sorts of things.

Yes, we should document this. We should also not use `MPI_COMM_WORLD` at all but instead let the user pass in whatever communicator they want -- not everyone works on `MPI_COMM_WORLD`, and if you don't this choice here will lead to a deadlock.